### PR TITLE
torch-based QMIX

### PR DESCRIPTION
**Pull request by YuechengLiu**
QMIX (based on torch)-- a value-based multi-agent algorithm.

### CODE REVIEW

File: benchmark/torch/qmix/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+QMIX is a **value-based** multi-agent reinforcement learning algorithm.  
+See more information about QMIX in: [QMIX: Monotonic Value Function Factorisation for Deep Multi-Agent Reinforcement Learning](https://arxiv.org/abs/1803.11485)
+### StarCraftII Environment
+Paper: [The StarCraft Multi-Agent Challenge](https://arxiv.org/pdf/1902.04043)  
+Github Repositories: [smac](https://github.com/oxwhirl/smac)  
+## Benchmark Results
+- We traineded our model in 5 different scenarios: *"3m", "8m", "2s_3z", "3s_5z"* and *"1c_3s_5z"*.  
+- The **difficulty** in all scenarios are set to be "7" (very difficult).  
+
+<img src="images/qmix_result.png" width = "1500"  alt="Performance" />
```
Comment (zenghsh3) [inline benchmark/torch/qmix/README.md]: The title of a picture should be below it.

Comment (zenghsh3) [inline benchmark/torch/qmix/README.md]: We should run at least three experiments for each environment.

Comment (zenghsh3) [inline benchmark/torch/qmix/README.md]: There are six environments in the paper, why don't we test in the six environments.

File: benchmark/torch/qmix/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+## How to Use
+### Dependencies
+- python
```
Comment (zenghsh3) [inline benchmark/torch/qmix/README.md]: [COMA](https://github.com/PaddlePaddle/PARL/tree/develop/benchmark/torch/coma) example requires python3.5+, Do you need it here?

File: benchmark/torch/qmix/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+## Start Training
+1. Modify the config in `train.py`.
+2. Start training:
+    ```bash
+    python train.py
+    ```
+3. View the training process with tensorboard:
+    ```bash
+    tensorboard --logdir ./
+    ```
```
Comment (zenghsh3) [inline benchmark/torch/qmix/README.md:32]: Maybe we should add the reference repo here.

File: benchmark/torch/qmix/qmix.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zenghsh3) [inline benchmark/torch/qmix/qmix.py]: We should move this file to the directory `parl/algorithms/torch`

File: benchmark/torch/qmix/qmix.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+
+class QMIX(parl.Algorithm):
+    def __init__(self,
+                 agent_model,
+                 qmixer_model,
+                 double_q=True,
+                 gamma=0.99,
+                 lr=0.0005,
+                 clip_grad_norm=None):
```
Comment (zenghsh3) [inline benchmark/torch/qmix/qmix.py]: Add comments of the arguments.

File: benchmark/torch/qmix/qmix.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+        device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
+        self.agent_model.to(device)
+        self.target_agent_model.to(device)
+        self.qmixer_model.to(device)
+        self.target_qmixer_model.to(device)
+
+        self.params = list(self.agent_model.parameters())
+        self.params += self.qmixer_model.parameters()
+        #self.optimizer = optim.Adam(self.params, lr=self.lr)
```
Comment (zenghsh3) [inline benchmark/torch/qmix/qmix.py]: remove this line.

File: benchmark/torch/qmix/qmixer_model.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class QMixerModel(nn.Module):
```
Comment (zenghsh3) [inline benchmark/torch/qmix/qmixer_model.py]: Why don't we inherit from `parl.Model`?

File: benchmark/torch/qmix/rnn_model.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class RNNModel(nn.Module):
```
Comment (zenghsh3) [inline benchmark/torch/qmix/rnn_model.py]: Why don't we inherit from parl.Model?

File: benchmark/torch/qmix/train.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+from qmix import QMIX
+from qmix_agent import QMixAgent
+import parl
+from parl.utils import logger
+from parl.utils import summary
+import numpy as np
+
+logger.set_dir('./log_path')
+
+config = {
```
Comment (zenghsh3) [inline benchmark/torch/qmix/train.py]: Maybe we can put the config to the file `qmix_config.py`

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: benchmark/torch/qmix/README.md (hunk 1 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/README.md]: "smac" doesn't provide "5m" in their Maps (scenarios).

Comment (YuechengLiu) [reply benchmark/torch/qmix/README.md]: updated

Comment (YuechengLiu) [reply benchmark/torch/qmix/README.md]: done.

File: benchmark/torch/qmix/README.md (hunk 2 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/README.md]: python>-3.6 is required to install pytorch==1.6.0. updated.

File: benchmark/torch/qmix/qmix.py (hunk 5 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/qmix.py]: modified

File: benchmark/torch/qmix/qmix.py (hunk 6 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/qmix.py]: updated

File: benchmark/torch/qmix/qmixer_model.py (hunk 7 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/qmixer_model.py]: updated

File: benchmark/torch/qmix/train.py (hunk 9 above)
Comment (YuechengLiu) [reply benchmark/torch/qmix/train.py]: modified.

File: benchmark/torch/qmix/README.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
+## QMIX based on Pytorch
+We reproduce the QMIX based on **PARL** and **Pytorch**, reaching the same level of indicators as the paper in StarCraftII benchmarks.
+### QMIX
+QMIX is a **value-based** multi-agent reinforcement learning algorithm.  
+See more information about QMIX in: [QMIX: Monotonic Value Function Factorisation for Deep Multi-Agent Reinforcement Learning](https://arxiv.org/abs/1803.11485)
+### StarCraftII Environment
+Paper: [The StarCraft Multi-Agent Challenge](https://arxiv.org/pdf/1902.04043)  
+Github Repositories: [smac](https://github.com/oxwhirl/smac)  
+## Benchmark Results
+<img src="images/torch-qmix-result.png" width = "500"  alt="Performance" />
```
Comment (zenghsh3) [inline benchmark/torch/qmix/README.md]: Please make these pictures in two rows.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

**Review by zenghsh3 — APPROVED** [review]
LGTM

**Pull request merged**